TWTValidation/JSON Validator/AST Nodes/TWTJSONSchemaTopLevelASTNode.m

```diff
@@ -40,4 +49,39 @@ - (void)acceptProcessor:(id<TWTJSONSchemaASTProcessor>)processor
     [processor processTopLevelNode:self];
 }
 
+
+- (NSArray *)allReferenceNodes
+{
+    // Assumes this only runs after the entire AST node tree has been generated, and no changes are made to the tree afterward
+    if (!_allReferenceNodes) {
+        _allReferenceNodes = [[self childrenReferenceNodes] copy];
+    }
+
+    return _allReferenceNodes;
+}
+
+
+- (TWTJSONSchemaASTNode *)nodeForReferenceNode:(TWTJSONSchemaReferenceASTNode *)referenceNode
+{
+    return [self nodeForPathComponents:[referenceNode.referencePathComponents mutableCopy]];
+}
+
+
+- (TWTJSONSchemaASTNode *)nodeForPathComponents:(NSMutableArray *)path
+{
+    // Assumes key is @"#"
+    if (path.count == 1) {
+        return self.schema;
+    }
+
+    [path removeObjectAtIndex:0];
+    return [self.schema nodeForPathComponents:path];
+}
```

Comment:
I’d prefer if this method took a normal array. It’s actually cheaper to use a subarrayWithRange:, and there’s nothing about this method that suggests that it should have a mutable array parameter.